drivers/gpu/drm/rp1/rp1-vec/rp1_vec.c

```diff
@@ -298,27 +298,43 @@ static enum drm_mode_status rp1vec_mode_valid(struct drm_device *dev,
 					      const struct drm_display_mode *mode)
 {
 	/*
-	 * Check the mode roughly matches one of our standard modes
-	 * (optionally half-height and progressive). Ignore H/V sync
-	 * timings which for interlaced TV are approximate at best.
+	 * Check the mode roughly matches something we can generate.
+	 * The hardware driver is very prescriptive about pixel clocks
+	 * and frame timings, but we'll tolerate some rounding errors.
+	 * It supports a range of frame sizes in each standard (which may be
+	 * useful to fine-tune overscan correction or emulate retro devices).
+	 * Ignore H/V sync timings which for interlaced TV are ill-defined.
 	 */
-	int i, prog;
 
-	prog = !(mode->flags & DRM_MODE_FLAG_INTERLACE);
-
-	for (i = 0; i < ARRAY_SIZE(rp1vec_modes); i++) {
-		const struct drm_display_mode *ref = rp1vec_modes + i;
-
-		if (mode->hdisplay == ref->hdisplay           &&
-		    mode->vdisplay == (ref->vdisplay >> prog) &&
-		    mode->clock + 2 >= ref->clock             &&
-		    mode->clock <= ref->clock + 2             &&
-		    mode->htotal + 2 >= ref->htotal           &&
-		    mode->htotal <= ref->htotal + 2           &&
-		    mode->vtotal + 2 >= (ref->vtotal >> prog) &&
-		    mode->vtotal <= (ref->vtotal >> prog) + 2)
-			return MODE_OK;
+	int prog = !(mode->flags & DRM_MODE_FLAG_INTERLACE);
+	int vtotal_full = mode->vtotal << prog;
+	int vdisplay_full = mode->vdisplay << prog;
+
+	/* Reject very small frames */
+	if (vtotal_full < 256 || mode->hdisplay < 256)
+		return MODE_BAD;
+
+	/* Check lines, frame period (ms) and vertical size limit */
+	if (vtotal_full >= 524 && vtotal_full <= 526 &&
+	    mode->htotal * vtotal_full > 33 * mode->clock &&
+	    mode->htotal * vtotal_full < 34 * mode->clock) {
+		if (vdisplay_full > 486)
+			return MODE_BAD;
+	} else if (vtotal_full >= 624 && vtotal_full <= 626 &&
+		   mode->htotal * vtotal_full > 39 * mode->clock &&
+		   mode->htotal * vtotal_full < 41 * mode->clock) {
+		if (vdisplay_full > 576)
+			return MODE_BAD;
+	} else {
+		return MODE_BAD;
 	}
+
+	/* Check pixel rate (kHz) and horizontal size limit */
+	if (mode->clock == 13500 && mode->hdisplay <= 720)
+		return MODE_OK;
+	if (mode->clock >= 15428 && mode->clock <= 15429 &&
+	    mode->hdisplay <= 800)
+		return MODE_OK;
```

Comment:
Minor nit that it would be nicer if `mode_valid` didn't switch from `if X return MODE_BAD` to `if X return MODE_OK` , and just had one `return MODE_OK` at the very end.

Reply:
Yes it bothered me as well but seemed nicer than the nested MODE_BADs in the vertical test. I'll see if I can tidy it.

Reply:
Now re-worked so the vertical test looks more like the horizontal one (both look for good cases).